rig-core/src/loaders/file.rs

```diff
@@ -0,0 +1,260 @@
+        FileLoader {
+            iterator: Box::new(self.iterator.map(|res| res.read())),
+        }
+    }
+    /// Reads the contents of the files within the iterator returned by `with_glob` or `with_dir`
+    ///  and returns the path along with the content.
+    ///
+    /// # Example
+    /// Read files in directory "files/*.txt" and print the content for cooresponding path for each
+    ///  file.
+    ///
+    /// ```rust
+    /// let content = FileLoader::with_glob("files/*.txt")?.read().into_iter();
+    /// for (path, result) in content {
+    ///     match result {
+    ///         Ok((path, content)) => println!("{:?} {}", path, content),
+    ///         Err(e) => eprintln!("Error reading file: {}", e),
+    ///     }
+    /// }
+    /// ```
+    pub fn read_with_path(self) -> FileLoader<'a, Result<(PathBuf, String), FileLoaderError>> {
+        FileLoader {
+            iterator: Box::new(self.iterator.map(|res| res.read_with_path())),
+        }
+    }
+}
+
+impl<'a, T: 'a> FileLoader<'a, Result<T, FileLoaderError>> {
+    /// Ignores errors in the iterator, returning only successful results. This can be used on any
+    ///  `FileLoader` state of iterator whose items are results.
+    ///
+    /// # Example
+    /// Read files in directory "files/*.txt" and ignore errors from unreadable files.
+    ///
+    /// ```rust
+    /// let content = FileLoader::with_glob("files/*.txt")?.read().ignore_errors().into_iter();
+    /// for result in content {
+    ///     println!("{}", content)
+    /// }
+    /// ```
+    pub fn ignore_errors(self) -> FileLoader<'a, T> {
+        FileLoader {
+            iterator: Box::new(self.iterator.filter_map(|res| res.ok())),
+        }
+    }
+}
+
+impl<'a> FileLoader<'a, Result<PathBuf, FileLoaderError>> {
+    /// Creates a new `FileLoader` using a glob pattern to match files.
+    ///
+    /// # Example
+    /// Create a `FileLoader` for all `.txt` files that match the glob "files/*.txt".
+    ///
+    /// ```rust
+    /// let loader = FileLoader::with_glob("files/*.txt")?;
+    /// ```
+    pub fn with_glob(
+        pattern: &str,
+    ) -> Result<FileLoader<Result<PathBuf, FileLoaderError>>, FileLoaderError> {
+        let paths = glob(pattern)?;
+        Ok(FileLoader {
+            iterator: Box::new(
+                paths
+                    .into_iter()
+                    .map(|path| path.map_err(FileLoaderError::GlobError)),
+            ),
+        })
+    }
+
+    /// Creates a new `FileLoader` on all files within a directory.
+    ///
+    /// # Example
+    /// Create a `FileLoader` for all files that are in the directory "files".
+    ///
+    /// ```rust
+    /// let loader = FileLoader::with_dir("files")?;
+    /// ```
+    pub fn with_dir(
+        directory: &str,
+    ) -> Result<FileLoader<Result<PathBuf, FileLoaderError>>, FileLoaderError> {
+        Ok(FileLoader {
+            iterator: Box::new(fs::read_dir(directory)?.map(|entry| Ok(entry?.path()))),
+        })
```

Comment:
Maybe we should pre-filter to only include paths that point to files (and not directories)?

Reply:
This applies to `PdfLoader` as well